java/ql/src/experimental/Security/CWE/CWE-522/InsecureBasicAuth.ql

```diff
@@ -0,0 +1,260 @@
+ */
+class BasicAuthString extends StringLiteral {
+  BasicAuthString() { exists(string s | this.getRepresentedString() = s | s.matches("Basic %")) }
+}
+
+/**
+ * String concatenated with `BasicAuthString`.
+ */
+predicate builtFromBasicAuthStringConcat(Expr expr) {
+  expr instanceof BasicAuthString
+  or
+  builtFromBasicAuthStringConcat(expr.(AddExpr).getLeftOperand())
+  or
+  exists(Expr other | builtFromBasicAuthStringConcat(other) |
+    exists(Variable var | var.getAnAssignedValue() = other and var.getAnAccess() = expr)
+  )
+}
+
+/** The `openConnection` method of Java URL. Not to include `openStream` since it won't be used in this query. */
+class HttpURLOpenMethod extends Method {
+  HttpURLOpenMethod() {
+    this.getDeclaringType() instanceof TypeUrl and
+    this.getName() = "openConnection"
+  }
+}
+
+/** Constructor of `ApacheHttpRequest` */
+predicate apacheHttpRequest(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(ConstructorCall cc |
+    cc.getConstructedType() instanceof ApacheHttpRequest and
+    node2.asExpr() = cc and
+    cc.getAnArgument() = node1.asExpr()
+  )
+}
+
+/** Constructors of `URI` */
+predicate createURI(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(URIConstructor cc |
+    node2.asExpr() = cc and
+    cc.getArgument(0) = node1.asExpr() and
+    cc.hasHttpStringArg()
+  )
+  or
+  exists(
+    StaticMethodAccess ma // URI.create
+  |
+    ma.getMethod().getDeclaringType().hasQualifiedName("java.net", "URI") and
+    ma.getMethod().hasName("create") and
+    node1.asExpr() = ma.getArgument(0) and
+    node2.asExpr() = ma
+  )
+}
+
+/** Constructors of `URL` */
+predicate createURL(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(URLConstructor cc |
+    node2.asExpr() = cc and
+    cc.getArgument(0) = node1.asExpr() and
+    cc.hasHttpStringArg()
+  )
+}
+
+/** Method call of `HttpURLOpenMethod` */
+predicate urlOpen(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(MethodAccess ma |
+    ma.getMethod() instanceof HttpURLOpenMethod and
+    node1.asExpr() = ma.getQualifier() and
+    ma = node2.asExpr()
+  )
+}
+
+/** Constructor of `BasicRequestLine` */
+predicate basicRequestLine(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(ConstructorCall mcc |
+    mcc.getConstructedType().hasQualifiedName("org.apache.http.message", "BasicRequestLine") and
+    mcc.getArgument(1) = node1.asExpr() and // `BasicRequestLine(String method, String uri, ProtocolVersion version)
+    node2.asExpr() = mcc
+  )
+}
+
+class BasicAuthFlowConfig extends TaintTracking::Configuration {
+  BasicAuthFlowConfig() { this = "InsecureBasicAuth::BasicAuthFlowConfig" }
+
+  override predicate isSource(DataFlow::Node src) {
+    src.asExpr() instanceof HttpString
+    or
+    exists(URLConstructor uc |
+      uc.hasHttpStringArg() and
+      src.asExpr() = uc.getArgument(0)
+    )
+    or
+    exists(URIConstructor uc |
+      uc.hasHttpStringArg() and
+      src.asExpr() = uc.getArgument(0)
+    )
```

Comment:
I think this causes duplicate findinds for the URI / URL constructors with a single string parameter, e.g.:
```java
String urlString = "http://google.com"; // One finding with this as flow start
URL url = new URL(urlString); // One finding with this as flow start
...
```
`hasHttpStringArg` should probably not match these constructors since their argument is a `HttpString` which would be matched here.
Related to https://github.com/github/codeql/pull/3976#discussion_r463951017.

Reply:
It seems that this isn't the root cause since it didn't make a difference after I made the code changes. I think the reason that both were reported is my implementation of `HttpString` - both the string literal `http://google.com` and the `url` variable were considered to be valid sources of the type `HttpString`. 

I've modified the implementation, now it's only reported once as shown in the expected results for the test case `testHttpUrlConnection` in [InsecureBasicAuth.java](https://github.com/github/codeql/blob/ff0dacf1d76a31970de4a4a6f2fa4c115d51f1cd/java/ql/test/experimental/query-tests/security/CWE-522/InsecureBasicAuth.java).